### DISCUSSION TITLE
Build multi-arch image

### DISCUSSION THREAD

**Discussion by xunholy** [fluxcd/flux2#303] [Proposals] [answered]
Although currently `arm64` & `amd64` architectures are supported it requires the `--arch` to alter and in a multi-arch cluster, this is also very limiting. GHCR supports multi-arch images.

I propose to move architecture-specific images and use `buildx` with `QEMU` to support more architectures. This would remove the need for `--arch` flag completely and allow for multi-architecture clusters without requiring to set affinity due to image limitations in their current state.

**Comment by stefanprodan** [comment]
We push both armv7 and arm64 to the same registry as we want to tract ARM usage vs AMD64. The gotk installer will pin the controllers on nodes based on the `--arch` argument so multi-architecture clusters are supported.

**Reply by xunholy** [reply]
What is the benefit of tracking architecture images, is there no API GHCR exposes with that type of information? It does seem to be a much easier solution to use a fat manifest rather than individual images per architectures.

**Reply by stefanprodan** [reply]
As I said both armv7 and amr64 are pushed to the same registry/image. And no GHCR doesn't offer stats per layer.

**Reply by xunholy** [reply]
> What is the benefit of tracking architecture images

Still curious...it seems like it could be simplified quite a lot but obviously, there is some benefit that has to outweigh it.

**Reply by hiddeco** [reply]
We are in a relatively early stage of development, and these numbers offer us maintainers a lot of value as they give us insights into our user base, which helps when we need to prioritize things.

ARM for example requires additional steps at times from a development perspective. If the percentage of ARM users is only a fraction of the total amount of AMD64 users, this means that these efforts may not be worth it at that given moment.

**Reply by onedr0p** [reply]
> ARM for example requires additional steps at times from a development perspective

@hiddeco do you care to elaborate on that, are you saying there is code written/conditioned specifically for ARM?

I've been building multiarch images w/ compiling Go for awhile now and I have never seen an issue.

**Reply by hiddeco** [reply]
See previous discussion: https://github.com/fluxcd/toolkit/discussions/188#discussioncomment-59318

**Reply by xunholy** [reply]
In all fairness that issue #188 seems systemic due to the same fundamental issue here which is the lack of multi-arch natively from the upstream maintainers which are why many people resort to self-maintained images until the upstream provides such support in their regular routine releases.

**Reply by hiddeco** [reply]
> We need to drop the kustomize binary and rewrite the controller parts that are shelling out to it.

**Reply by onedr0p** [reply]
Thanks for linking, but GHCR supports multiarch images now, and grabbing external binaries is pretty easy as mentioned using `TARGETPLATFORM`. See [here](https://github.com/raspbernetes/multi-arch-images/blob/master/build/flux/Dockerfile#L44) how we do it for Flux. Basically you can grab the arch for the kubectl binary using logic like `GOARCH=$(echo ${TARGETPLATFORM} | cut -d / -f2)`. I agree this would take some testing to make sure you are pulling in the right external binaries, but in the long run it would pay off.

**Reply by onedr0p** [reply]
Also I did see a message from a GitHub employee that said GHCR isn't meant for production images (yet?). Maybe that is also worth thinking about? See response from clarkbw in https://news.ycombinator.com/item?id=24346645

> We recommend using GHCR for the development and test workflows and then publishing to the "Cloud *CRs" for your production images such that they can be pulled directly from there.

**Reply by hiddeco** [reply]
1. `kustomize` (the binary) does not have support for ARM64, so `TARGETPLATFORM` would not have brought you anywhere.
2. We want to track ARM vs AMD64 usage. Which at present is not possible without providing separate images.

This is discussion is starting to show signs of a vicious circle instead of being fruitful.

**Reply by onedr0p** [reply]
To be fair I'm okay with you guys not supporting multiarch images, I'm just a terrible debater 😅 I apologize for that.

**Reply by xunholy** [reply]
To keep things constructive @hiddeco with `buildx` and the appropriate approach we can help support the new experience with multi-arch images, it might mean losing that valuable piece of information but for a far better UX. Things like `kustomize` aren't supported because their upstream don't provide it, there are plenty of people already using it with ARM64 in many environments with almost no changes by leveraging `buildx`. EG. https://github.com/raspbernetes/multi-arch-images/blob/master/build/kustomize/Dockerfile

**Reply by stefanprodan** [reply]
We do publish a multi-arch image but only for ARM (armv7+arm64) and we do use buildx. `docker pull ghcr.io/fluxcd/kustomize-controller-arm64:v0.1.0` works on armv7/v8. We have no intention of merging the ARM layers under the main image.

**Reply by onedr0p** [reply]
One last question, wouldn't there be an issue where if you are running a multi-arch cluster and you deployed gotk with the `--arch=arm64` flag, now imagine a scenario where you need to take your arm nodes offline for whatever reason. The deployments would stop working, correct?

The fix it seems would be to rerun gotk bootstrap without the `--arch` flag, but that requires manual intervention.

**Comment by onedr0p** [comment]
I agree with @xUnholy , having one image would simplify builds and ease of use.

**Reply by stefanprodan** [reply]
What builds are those?

**Reply by onedr0p** [reply]
The CI pipelines, I've been building multiarch images for awhile now and the process is pretty streamlined.

**Reply by stefanprodan** [reply]
You don't have to build GOTK we do that for you.

**Reply by onedr0p** [reply]
I know, I'm talking about your own maintaince of building images and reducing lines of code, there and in gotk source :)

But if you want to really track usage of separate arch, there's no way to do that with a multiarch image.

**Comment by billimek** [comment]
Multi-arch images would be useful for kubernetes clusters operating in mixed-architecture environments.  Having separate arch-specific images or tags doesn't allow for such operation.

**Reply by onedr0p** [reply]
@billimek as mentioned by @stefanprodan this is a non-issue since the gotk deployment controllers are attracted to whatever you pass into the `--arch` flag. I do see this as being confusing at first and not great in terms of UX. Removing the need for this flag and building multi-arch images would be 10x better for UX.

**Comment by stefanprodan** [comment, marked as answer]
To sum this up as I feel it is going nowhere:
- we do support the following architectures: AMD64, ARM64, ARMv7
- we do publish multi-arch images for both ARM64 and ARMv7 under `ghcr.io/fluxcd/<controller>-arm64:semver`
- the gotk installer works on mixed-architecture clusters, you need to specify the arch you want to pin the controllers to with `gotk install --arch=amd64|arm64|arm`
- we want to track ARM usage so we have no intention of merging all layers under a single image
